Python/pylifecycle.c

```diff
@@ -1348,6 +1351,38 @@ Py_FinalizeEx(void)
     PyThreadState *tstate = _PyRuntimeState_GetThreadState(runtime);
     PyInterpreterState *interp = tstate->interp;
 
+    /* Check we're running in the main interpreter (not yet supported to call
+     * from any interpreter).
+     */
+    if (interp != PyInterpreterState_Main()) {
+        fprintf(stderr,
+                "Py_FinalizeEx: error: must be called from the main interpreter\n");
+        return -1;
+    }
+
+    // Finalize sub-interpreters.
+    runtime->interpreters.allow_new = 0;
```

Comment:
It may be safer to acquire _PyRuntime.interpreters.mutex beforing setting this variable. It may be better to move this code into pystate.c, since this file control the list of interpreters.

Reply:
I presume you're suggesting refactoring the 'finalize subinterpreters' logic? It would seem to me a function that does 'finalizing' belongs better in pylifecycle than pystate?

I've added in an acquisition of the lock here for now.